DOC: draft of release notes (#737)
bluesky/ophyd: doc/source/release_notes.rst

 Release History
=================

v1.4.0 (2019-06-XX)
===================

Features
--------

> * Version ohpyd classes to support IOCs changing over time. This is currently

Reviewer: ```suggestion
* Version ophyd classes to support IOCs changing over time. This is currently
```

>   implemented from AreaDetector and it's plugins (support for 1-9 to 3-5).

Reviewer: ```suggestion
  implemented for AreaDetector plugins (releases 1-9 to 3-4).
```

> * `.NDDerivedSignal` to expose a re-shaped version of an N-D array.

Reviewer: ```suggestion
* Added `NDDerivedSignal`, which supplies a reshaped version of an input array signal.
```

> * `~.ophyd.device.Component.subscriptions` decorator to set up default

Reviewer: ```suggestion
* Added `~.ophyd.device.Component.subscriptions` decorator to set up default
```

  subscriptions on a component.
* Added `.create_device_from_components` to aid dynamic device generation.
* Added `.required_for_connection` decorator to mark the Components that must
  be connected for the overall Device to be considered connected.
* Added a hook to be called on all ophyd object creation via the

>   `.ohpydobj.add_instantiation_callback` class method along with the

Reviewer: ```suggestion
  `.ophydobj.add_instantiation_callback` class method along with the
```

  `.register_instances_keyed_on_name` and
  `.register_instance_in_weakset` helper functions.
* Added `.ophydobj.dotted_name` property to get the full attribute
  name of a child component.
* Added ``read_access`` and ``write_access`` properties.

> * Added ``destroy`` method to all ohpyd objects.

Reviewer: ```suggestion
* Added ``destroy`` method to all ophyd objects.
```

API Changes
-----------

* Drop support for Python3.5

> * If using pyepics, requires pyepics >= 3.4.0

Reviewer: ```suggestion
* The optional dependency pyepics, if installed, must be above version 3.4.0 to be used.
```

* If you are holding a reference to a pyepics.pv.PV that is shared
  with ophyd and all ophyd objects that use that PV are torn down, all
  callbacks on the PV will be cleared and the channel will be torn
  down.  If this is a problem for you, please create a bug report.
* Removed ``ophyd.control_layer``.  The "control layer" used to access
  EPICS can be controlled via `~ophyd.set_cl`.
* `.DynamicDeviceComponent` is now a `.Component` sub-class

> * Change argument name in `.ophydobj.subscribe` from ``cb`` to

Reviewer: ```suggestion
* Changed argument name in `.ophydobj.subscribe` from ``cb`` to
```

  ``callback``.
* Removed ``ophyd.tests.conftest.FakeEpicsPV``,

>   ``ophyd.tests.conftest.FakeEpicsWaveForm`` , and associated helper

Reviewer 2: ```suggestion
  ``ophyd.tests.conftest.FakeEpicsWaveForm``, and associated helper
```

  functions.  If you need this class, please vendor it from a previous

>   version of ophyd.

Reviewer: ```suggestion
  version of ophyd - or please consider moving to `make_fake_device` or `caproto`-based simulation IOCs.
```

> * Removed ``ophyd.tests.AssertTolls`` and use standard pytest idioms

Reviewer: ```suggestion
* Removed ``ophyd.tests.AssertTools`` and use standard pytest idioms
```

>   version of ophyd - or please consider moving to `make_fake_device` or `caproto`-based simulation IOCs.

Reviewer 2: ```suggestion
  version of ophyd - or please consider moving to `make_fake_device`
  or `caproto`-based simulation IOCs.
```

>   throughout test suite

Reviewer 2: ```suggestion
  throughout test suite.
```

Internals
---------
* switch from `.Device` using meta-classes to using a

>   ``__init__subclass__``.  There are now no meta-classes in ohpyd!

Reviewer: ```suggestion
  ``__init__subclass__``.  There are now no meta-classes in ophyd!
```

> * switch from `.Device` using meta-classes to using a

Reviewer 2: ```suggestion
* Switch from `.Device` using meta-classes to using a
```

* Completely overhauled how we connect to PVs at initialization and tear
  them down on destruction of the ophyd objects.
* Completely overhauled how PV meta-data is handled.
* Completely overhauled the dispatcher thread mechanism.

> * Removed our backport of enums.

Reviewer: ```suggestion
* Removed our backport of `enum`, as it is available on the minimum Python 3.6.
```

> 

Reviewer 2: Do we need that many blank spaces?

Reviewer: Definitely not!
